contracts/src/base/registry/facets/distribution/v2/StakingRewards.sol

```diff
@@ -0,0 +1,329 @@
+      $.treasureByBeneficiary[oldBeneficiary].earningPower -= amount;
+    }
+
+    updateReward($, newBeneficiary);
+    deposit.beneficiary = newBeneficiary;
+    unchecked {
+      // the invariant totalStaked >= treasureByBeneficiary[beneficiary].earningPower is ensured on stake and increaseStake
+      // the following won't overflow
+      $.treasureByBeneficiary[newBeneficiary].earningPower += amount;
+    }
+    // TODO: emit events
+  }
+
+  function withdraw(
+    Layout storage $,
+    uint256 depositId,
+    uint96 amount
+  ) internal {
+    updateGlobalReward($);
+    Deposit storage deposit = $.deposits[depositId];
+    // cache storage reads
+    (address beneficiary, address owner) = (deposit.beneficiary, deposit.owner);
+    updateReward($, beneficiary);
+
+    deposit.amount -= amount;
+    unchecked {
+      // totalStaked >= deposit.amount
+      $.totalStaked -= amount;
+      // stakedByDepositor[owner] >= deposit.amount
+      $.stakedByDepositor[owner] -= amount;
+      // treasureByBeneficiary[beneficiary].earningPower >= deposit.amount
+      $.treasureByBeneficiary[beneficiary].earningPower -= amount;
+    }
+    $.stakeToken.safeTransferFrom(
+      $.delegationMinions[deposit.delegatee],
+      owner,
+      amount
+    );
+    // TODO: emit events
+  }
+
+  function claimReward(
+    Layout storage $,
+    address beneficiary
+  ) internal returns (uint256 reward) {
+    updateGlobalReward($);
+    updateReward($, beneficiary);
+
+    Treasure storage treasure = $.treasureByBeneficiary[beneficiary];
+    reward = treasure.unclaimedRewardSnapshot / SCALE_FACTOR;
+    if (reward != 0) {
+      unchecked {
+        treasure.unclaimedRewardSnapshot -= reward * SCALE_FACTOR;
+      }
+      $.rewardToken.safeTransfer(beneficiary, reward);
+      // TODO: emit events
+    }
+  }
+
+  function notifyRewardAmount(Layout storage $, uint256 reward) internal {
+    if (!$.isRewardNotifier[msg.sender])
+      StakingRewards_InvalidRewardNotifier.selector.revertWith();
+
+    $.rewardPerTokenAccumulated = currentRewardPerTokenAccumulated($);
+
+    // cache storage reads
+    (uint256 rewardDuration, uint256 rewardEndTime) = (
+      $.rewardDuration,
+      $.rewardEndTime
+    );
+
+    uint256 rewardRate;
+    if (block.timestamp >= rewardEndTime) {
+      rewardRate = reward.mulDiv(SCALE_FACTOR, rewardDuration);
+    } else {
+      uint256 remainingTime;
+      unchecked {
+        remainingTime = rewardEndTime - block.timestamp;
+      }
+      uint256 leftover = $.rewardRate * remainingTime;
+      rewardRate = (leftover + reward * SCALE_FACTOR) / rewardDuration;
+    }
+
+    // batch storage writes
+    ($.rewardEndTime, $.lastUpdateTime, $.rewardRate) = (
+      block.timestamp + rewardDuration,
+      block.timestamp,
+      rewardRate
+    );
+
+    if (rewardRate < SCALE_FACTOR)
+      StakingRewards_InvalidRewardRate.selector.revertWith();
```

Comment:
why not just `revert StakingRewards_InvalidRewardRate()`

Reply:
Every `if (condition) revert Something();` creates a new [`JUMPDEST`](https://www.evm.codes/#5b?fork=cancun) and additional bytecode which increases the contract size. Using a unified internal function to revert in the scratch space solves this.